File: 树/331. Verify Preorder Serialization of a Binary Tree.js
/**
 * @param {string} preorder
 * @return {boolean}
 */
var isValidSerialization = function(preorder) {
    const tree = preorder.split(',');
    let idx = 0;

    const dfs = () => {
        if (idx >= tree.length) {
            return false;
        }

        if (tree[idx] === '#') {
            idx++;
            return true;
        } else {
            idx++;
            return dfs() && dfs();
        }
    };

    return dfs() && idx === tree.length;
};

File: 树/331. Verify Preorder Serialization of a Binary Tree.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        tree = split(preorder);
        return dfs() && idx == tree.size();;
    }
private:
    vector<string> tree;
    int idx;
    
    bool dfs() {
        if (idx >= tree.size()) return false;
        const string& cur = tree[idx++];
        if (cur == "#") return true;
        
        return dfs() && dfs();
    }
    
    vector<string> split(const string& str) {
        string tmp;
        vector<string> ans;
        for (const char ch : str) {
            if (ch == ',') {
                ans.push_back(tmp);
                tmp.clear();
            } else {
                tmp += ch;
            }
        }
        
        if (!tmp.empty()) ans.push_back(tmp);
        return ans;
    }
};